### PR TITLE
Rule and e2e changes for unified config model

### PR DESCRIPTION
**Pull request by frankbu**
**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by kyessenov** [comment]
Thanks, this is great. Let's keep it open until we make changes to the api and code to reflect the new design.

**Comment by linsun** [comment]
Very nice Frank!  Will we give any useful warning if user tries to do `kubectl apply -f` on the v0.1 route rule yaml file?  I assume an error will be thrown.

**Comment by frankbu** [comment]
@kyessenov I agree. I think of this as a living document to keep track of the breaking changes that we're making. I'll try to keep it in sync with any design changes that might still be forthcoming. Maybe we don't ever merge this particular PR, but instead move the final contents to some 0.1->0.2 migration doc.

@linsun not sure how much validation we'll have in 0.2, but hopefully any errors will be clear enough. In general, since we are in very early point releases, I don't think we need to worry too much about the breaking changes as long as we highlight very clearly in the release notes for 0.2 that the rule format has changed, with no backward compat, so users must convert their rules as documented here.

**Comment by yutongz** [comment]
Could you add release-note message in pr description like: 
``release-note
XXX or NONE
``

**Comment by frankbu** [comment]
Updated the e2e tests to use the new route-rule format and to create the rules using `kubectl` instead of `istioctl`. New version of the rule yaml files are in a `rules` subdirectory of the bookinfo sample dir (to keep the old ones around for now, and also cleaner IMO). Once they're working, we can delete the old files. Mixer rules are currently unchanged.

Using `v1alpha2` for the resource version for now. @kyessenov Let me know if you think it should be `v1beta1` instead.

This PR is dependent on https://github.com/istio/pilot/issues/1138.

File: samples/apps/bookinfo/rules/destination-policy-reviews.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+apiVersion: config.istio.io/v1beta1
+kind: DestinationPolicy
+metadata:
+  name: reviews
+spec:
+  policy:
```
Comment (kyessenov) [inline samples/apps/bookinfo/rules/destination-policy-reviews.yaml]: Let's use the new format and deprecate the policy list. The code is already implemented in pilot PR.

**Comment by frankbu** [comment]
Kuat, I tired to run kubeTest as my first test:
```
./kubeTest.sh -m "gcr.io/istio-testing,7b4b12d31b70fbae9001139f095ae47ed0a48185" -i https://storage.googleapis.com/istio-artifacts/pilot/7b4b12d31b70fbae9001139f095ae47ed0a48185/artifacts/istioctl -s
```

It fails at the first route-rule create with messages like this:
```
I0901 14:00:52.272290   15247 request.go:559] Request Body: "{\"apiVersion\":\"config.istio.io/v1alpha2\",\"kind\":\"RouteRule\",\"metadata\":{\"name\":\"productpage-default\",\"namespace\":\"bookinfo-test-55253859\"},\"spec\":{\"destination\":{\"name\":\"productpage\"},\"precedence\":1,\"route\":[{\"labels\":{\"version\":\"v1\"}}]}}\n"
I0901 14:00:52.272342   15247 round_trippers.go:299] curl -k -v -XPOST  -H "Content-Type: application/json" -H "User-Agent: kubectl/v1.5.2 (linux/amd64) kubernetes/08e0995" -H "Accept: application/json" https://192.168.99.101:8443/apis/config.istio.io/v1alpha2/namespaces/bookinfo-test-55253859/routerules
I0901 14:00:52.273388   15247 round_trippers.go:318] POST https://192.168.99.101:8443/apis/config.istio.io/v1alpha2/namespaces/bookinfo-test-55253859/routerules 404 Not Found in 1 milliseconds
I0901 14:00:52.273406   15247 round_trippers.go:324] Response Headers:
I0901 14:00:52.273410   15247 round_trippers.go:327]     X-Content-Type-Options: nosniff
I0901 14:00:52.273412   15247 round_trippers.go:327]     Content-Length: 19
I0901 14:00:52.273415   15247 round_trippers.go:327]     Date: Fri, 01 Sep 2017 14:00:53 GMT
I0901 14:00:52.273418   15247 round_trippers.go:327]     Content-Type: text/plain; charset=utf-8
I0901 14:00:52.273657   15247 request.go:904] Response Body: 404 page not found
I0901 14:00:52.273681   15247 request.go:995] Response Body: "404 page not found\n"
```

I tried to create a rule with kubectl, and got similar response:
```
Franks-MacBook-Pro:istio frankbudinsky$ kubectl apply -f samples/apps/bookinfo/rules/route-rule-reviews-v3.yaml -n bookinfo-test-55253859
Error from server (NotFound): error when creating "samples/apps/bookinfo/rules/route-rule-reviews-v3.yaml": the server could not find the requested resource (post routerules.config.istio.io)
```

Am I missing some setup or wrong version? I'm using minikube with kube 1.7
```
Franks-MacBook-Pro:istio frankbudinsky$ kubectl version
Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.3", GitCommit:"2c2fe6e8278a5db2d15a013987b53968c743f2a1", GitTreeState:"clean", BuildDate:"2017-08-03T07:00:21Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.0", GitCommit:"d3ada0119e776222f11ec7945e6d860061339aad", GitTreeState:"clean", BuildDate:"2017-07-26T00:12:31Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

File: samples/apps/bookinfo/rules/destination-policy-reviews.yaml (hunk 1 above)
Comment (frankbu) [reply samples/apps/bookinfo/rules/destination-policy-reviews.yaml]: OK, I'll change it

**Comment by frankbu** [comment]
Maybe there's a disconnect on what api to use. I've changed the tests to use `kubectl` to create rules, but I noticed that in istio/pilot#1160, `istioctl` is used. I tried istioctl, and that seems to work:
```
ubuntu@ubuntu-xenial:~/go/src/istio.io/pilot/.idea-istio-bookinfo-project/istio$ /tmp/istioctl create -n bookinfo-test-55253859 -f samples/apps/bookinfo/rules/route-rule-reviews-v3.yaml
Created config route-rule/bookinfo-test-55253859/reviews-default at revision 383122
ubuntu@ubuntu-xenial:~/go/src/istio.io/pilot/.idea-istio-bookinfo-project/istio$ /tmp/istioctl get route-rule -n bookinfo-test-55253859
route-rule/bookinfo-test-55253859/reviews-default
```
but, this changes the rule format from this:
```
apiVersion: config.istio.io/v1alpha2
kind: RouteRule
metadata:
  name: reviews-default
spec:
  destination:
    name: reviews
  precedence: 1
  route:
  - labels:
      version: v3
    weight: 100
```
to this:
```
type: route-rule
name: reviews-default
spec:
  destination:
    name: reviews
  precedence: 1
  route:
  - labels:
      version: v3
    weight: 100
```

Is the plan to stick with istioctl for rules? If so, I need to change the tests back.

**Comment by kyessenov** [comment]
Let's stick with kubectl in the test infrastructure instead of istioctl. That is one less dependency.
This problem you described is due to the issue with dashes that I have a fix for, but is hard to land as it affects the clusters in a non-backwards compatible way.

**Comment by frankbu** [comment]
Kuat, I ran kubeTest.sh again with your latest fix and it gets much farther :-)

Now it seems to only fail when the test tries to replace a rule (`reviews-default`) with a different one (https://github.com/istio/istio/blob/4d557f38d643917ce4ffd5d74c3f12f1684565ea/tests/kubeTest.sh#L200). 

Here's the output:
```
I0901 21:51:51.129432   16500 round_trippers.go:299] curl -k -v -XPUT  -H "Accept: application/json" -H "Content-Type: application/json" -H "User-Agent: kubectl/v1.5.2 (linux/amd64) kubernetes/08e0995" https://192.168.99.101:8443/apis/config.istio.io/v1alpha2/namespaces/bookinfo-test-276a7789/routerules/reviews-default
I0901 21:51:51.201289   16500 round_trippers.go:318] PUT https://192.168.99.101:8443/apis/config.istio.io/v1alpha2/namespaces/bookinfo-test-276a7789/routerules/reviews-default 404 Not Found in 71 milliseconds
I0901 21:51:51.201313   16500 round_trippers.go:324] Response Headers:
I0901 21:51:51.201316   16500 round_trippers.go:327]     Content-Type: application/json
I0901 21:51:51.201319   16500 round_trippers.go:327]     Content-Length: 252
I0901 21:51:51.201322   16500 round_trippers.go:327]     Date: Fri, 01 Sep 2017 21:51:51 GMT
I0901 21:51:51.201367   16500 request.go:904] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"routerules.config.istio.io \"reviews-default\" not found","reason":"NotFound","details":{"name":"reviews-default","group":"config.istio.io","kind":"routerules"},"code":404}
I0901 21:51:51.203029   16500 request.go:995] Response Body: "{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"routerules.config.istio.io \\\"reviews-default\\\" not found\",\"reason\":\"NotFound\",\"details\":{\"name\":\"reviews-default\",\"group\":\"config.istio.io\",\"kind\":\"routerules\"},\"code\":404}\n"
I0901 21:51:51.203889   16500 helpers.go:203] server response object: [{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "error when replacing \"/home/ubuntu/go/src/istio.io/pilot/.idea-istio-bookinfo-project/istio/samples/apps/bookinfo/rules/route-rule-reviews-50-v3.yaml\": routerules.config.istio.io \"reviews-default\" not found",
  "reason": "NotFound",
  "details": {
    "name": "reviews-default",
    "group": "config.istio.io",
    "kind": "routerules"
  },
  "code": 404
}]
F0901 21:51:51.203915   16500 helpers.go:116] Error from server (NotFound): error when replacing "/home/ubuntu/go/src/istio.io/pilot/.idea-istio-bookinfo-project/istio/samples/apps/bookinfo/rules/route-rule-reviews-50-v3.yaml": routerules.config.istio.io "reviews-default" not found
/home/ubuntu/go/src/istio.io/pilot/.idea-istio-bookinfo-project/istio/tests/istioUtils.sh: line 30: Could not replace rule (exit 1)
```

Maybe PUT needs the same fix as POST?

**Comment by kyessenov** [comment]
kubectl replace requires that the rule with the same name is already present. Seems like that's not true here.

**Comment by frankbu** [comment]
It's definitely set in the earlier part of the test where the default routes are set (https://github.com/istio/istio/blob/4d557f38d643917ce4ffd5d74c3f12f1684565ea/samples/apps/bookinfo/rules/route-rule-all-v1.yaml). That test passes, and we are now much farther along in the test (almost the end, actually).

I did notice that `kubectl get routerules` doesn't return anything either
```
$ kubectl get routerule -n bookinfo-test-276a7789
No resources found.
```

Maybe GET is the problem?

**Comment by kyessenov** [comment]
closing since superseded by another PR

**Pull request closed without merging**